src/openms/include/OpenMS/FILTERING/ID/IDFilter.h

```diff
@@ -1002,6 +1106,204 @@ namespace OpenMS
+      keepMatchingPeptideHits(prot_and_pep_ids, best_per_peptide);
+    }
+
+    static void annotateBestPerPeptidePerRun(const std::vector<ProteinIdentification>& prot_ids, std::vector<PeptideIdentification>& pep_ids, bool ignore_mods, bool ignore_charges, Size nr_best_spectrum)
+    {
+      std::map<std::string, std::unordered_map<std::string, std::map<Int, PeptideHit*>>> best_peps_per_run;
+      for (const auto& id : prot_ids)
+      {
+        best_peps_per_run[id.getIdentifier()] = std::unordered_map<std::string, std::map<Int, PeptideHit*>>();
+      }
+      annotateBestPerPeptidePerRunWithData(best_peps_per_run, pep_ids, ignore_mods, ignore_charges, nr_best_spectrum);
+    }
+
+    /// Annotates PeptideHits from PeptideIdentification if it is the best peptide hit for its peptide sequence
+    /// Adds metavalue "bestForItsPeps" which can be used for additional filtering.
+    static void annotateBestPerPeptidePerRunWithData(std::map<std::string, std::unordered_map<std::string, std::map<Int, PeptideHit*>>>& best_peps_per_run, std::vector<PeptideIdentification>& pep_ids, bool ignore_mods, bool ignore_charges, Size nr_best_spectrum)
+    {
+      for (auto &pep : pep_ids)
+      {
+        //skip if no hits (which almost could be considered and error or warning.
+        if (pep.getHits().empty())
+          continue;
+
+        bool higher_score_better = pep.isHigherScoreBetter();
+        //make sure that first = best hit
+        pep.sort();
+
+        auto& best_pep_map = best_peps_per_run[pep.getIdentifier()];
+
+        auto pepIt = pep.getHits().begin();
+        auto pepItEnd = nr_best_spectrum == 0 || pep.getHits().size() <= nr_best_spectrum ? pep.getHits().end() : pep.getHits().begin() + nr_best_spectrum;
+        for (; pepIt != pepItEnd; ++pepIt)
+        {
+          PeptideHit &hit = *pepIt;
+
+          String lookup_seq;
+          if (ignore_mods)
+          {
+            lookup_seq = hit.getSequence().toUnmodifiedString();
+          }
+          else
+          {
+            lookup_seq = hit.getSequence().toString();
+          }
+
+          int lookup_charge = 0;
+          if (!ignore_charges)
+          {
+            lookup_charge = hit.getCharge();
+          }
+
+
+          auto it_inserted = best_pep_map.emplace(std::move(lookup_seq), std::map<Int, PeptideHit*>());
+          auto it_inserted_chg = it_inserted.first->second.emplace(lookup_charge, &hit);
+          PeptideHit* &p = it_inserted_chg.first->second; //either the old one if already present, or this
+          if (!it_inserted_chg.second) //was already present -> possibly update
+          {
+            if (
+                (higher_score_better && (hit.getScore() > p->getScore())) ||
+                (!higher_score_better && (hit.getScore() < p->getScore()))
+                )
+            {
+              p->setMetaValue("bestForItsPep", "false");
+              hit.setMetaValue("bestForItsPep", "true");
```

Comment:
@timosachsenberg cant we use `true` or `false` as meta values? do we need the string here?

Reply:
PS: if you use this only internally and not in a featureXML or anwywhere, maybe use 0 / 1 integer values and see if that is faster.

Reply:
I think this would require larger changes to DataValue.
I hope small string optimiation is kicking in -  in that case it takes the same amount of space.

Reply:
True. I will switch to 0 and 1.